packages/ai-jsx/src/core/conversation.tsx

```diff
@@ -309,3 +323,204 @@ export async function* ShowConversation(
+ * effect within the messages themselves.
+ */
+export async function ShrinkConversation(
+  {
+    cost: costFn,
+    budget: budget,
+    children,
+  }: {
+    cost: (message: ConversationMessage, render: AI.RenderContext['render']) => Promise<number>;
+    budget: number;
+    children: Node;
+  },
+  { render, memo, logger }: AI.ComponentContext
+) {
+  /**
+   * We construct a tree of immutable and shrinkable nodes such that shrinkable nodes
+   * can contain other nodes.
+   */
+  type TreeNode = ImmutableTreeNode | ShrinkableTreeNode;
+
+  interface ImmutableTreeNode {
+    type: 'immutable';
+    element: AI.Element<any>;
+    cost: number;
+  }
+
+  interface ShrinkableTreeNode {
+    type: 'shrinkable';
+    element: AI.Element<AI.PropsOfComponent<typeof InternalShrinkable>>;
+    cost: number;
+    children: TreeNode[];
+  }
+
+  /** Converts a conversational `AI.Node` into a shrinkable tree. */
+  async function conversationToTreeRoots(conversation: AI.Node): Promise<TreeNode[]> {
+    const rendered = await render(conversation, {
+      stop: (e) => isConversationalComponent(e) || e.tag === InternalShrinkable,
+    });
+
+    const asTreeNodes = await Promise.all(
+      rendered.map<Promise<TreeNode | null>>(async (value) => {
+        if (typeof value === 'string') {
+          return null;
+        }
+
+        if (value.tag === InternalShrinkable) {
+          const children = await conversationToTreeRoots(value.props.children);
+          return { type: 'shrinkable', element: value, cost: aggregateCost(children), children };
+        }
+
+        return {
+          type: 'immutable',
+          element: value,
+          cost: await costFn(toConversationMessages([value])[0], render),
+        };
+      })
+    );
+
+    return asTreeNodes.filter((n): n is TreeNode => n !== null);
+  }
+
+  /** Finds the least important node in the tree, considering cost as a second factor. */
+  function leastImportantNode(roots: TreeNode[]): ShrinkableTreeNode | undefined {
+    function compareImportance(nodeA: ShrinkableTreeNode, nodeB: ShrinkableTreeNode) {
+      // If the two nodes are of the same importance, consider the higher cost node less important.
+      return nodeA.element.props.importance - nodeB.element.props.importance || nodeB.cost - nodeA.cost;
+    }
+
+    let current = undefined as ShrinkableTreeNode | undefined;
+    roots.forEach((node) => {
+      if (node.type !== 'shrinkable') {
+        return;
+      }
+
+      if (current === undefined || compareImportance(node, current) < 0) {
+        current = node;
+      }
+
+      const leastImportantDescendant = leastImportantNode(node.children);
+      if (leastImportantDescendant !== undefined && compareImportance(leastImportantDescendant, current) < 0) {
+        current = leastImportantDescendant;
+      }
+    });
+
+    return current;
+  }
+
+  function aggregateCost(roots: TreeNode[]): number {
+    return roots.reduce((cost, node) => cost + node.cost, 0);
```

Comment:
```suggestion
    return _.sumBy(roots, 'cost')
```

I think this will work